module/zfs/zvol.c

```diff
@@ -966,13 +968,22 @@ zvol_request(struct request_queue *q, struct bio *bio)
 		 */
 		zvr->rl = zfs_range_lock(&zv->zv_range_lock, offset, size,
 		    RL_WRITER);
+		/*
+		 * FIXME: if there is at least another lock held on the same
+		 * block process this request synchronously.
+		 */
+		if (zfs_range_locked(&zv->zv_range_lock, P2ALIGN(offset,
+		    zv->zv_volblocksize), zv->zv_volblocksize) > 1)
+			need_sync = B_TRUE;
```

Comment:
@behlendorf sorry this is not completely clear to me: if `zil_commit()` is called directly by `zvol_write()` and we need to avoid running from a taskq thread i think we need to do it here in `zvol_request()`?